### PR TITLE
feature: track and hold onto best eval model

### PR DESCRIPTION
**Pull request by kallewoof**
With model checkpoints being trimmed away, it is possible that one might auto-lose the best evaluating checkpoint in the current set up. This PR compares the eval loss to the best seen and saves a "best_loss" checkpoint that is overwritten each time a new best is achieved.

1. I'm not sure I am getting the loss for all evals now. I think I do, but it would be great to have this confirmed.
2. Is there a more elegant way to relay `pending_best_loss_save` than doing a file write and `os.path.exist` check in the process_step loop? My deepspeed knowledge is limited.

### CODE REVIEW

File: saver.py [hunk 1]
```diff
@@ -181,5 +189,32 @@ def process_step(self, step):
...
+        if loss is not None:
+            if self.best_loss is None:
+                print(f"Evaluation loss: {loss}")
+            elif loss >= self.best_loss:
+                print(f"Evaluation loss: {loss} (best: {self.best_loss}, Δ: {self.best_loss - loss} = {100 * (self.best_loss - loss) / self.best_loss:.2f}%)")
+            if self.best_loss is None or loss < self.best_loss:
+                self.old_best = self.best_loss
+                self.best_loss = loss
+                if save_best:
+                    with open(".pending_save_best_loss", "w") as f:
```
Comment (tdrussell) [inline saver.py:218]: This file should probably be inside save_root. If you killed the job at the right time, it could persist before being renamed. Then if you launched another training run it might do weird things.

File: saver.py [hunk 2]
```diff
@@ -181,5 +189,32 @@ def process_step(self, step):
...
 
+        pending_save_best_loss = os.path.exists(".pending_save_best_loss")
+        if pending_save_best_loss:
+            self.save_model('best_loss')
+            if is_main_process():
+                if self.old_best is not None:
+                    print(f'New best loss: {self.best_loss} from {self.old_best} (Δ{self.old_best - self.best_loss} = {100 * (self.old_best - self.best_loss) / self.old_best:.2f}%)')
+                else:
+                    print(f'New best loss: {self.best_loss}')
+                if os.path.exists(os.path.join(self.save_root, 'best_loss.txt')):
```
Comment (tdrussell) [inline saver.py:200]: I think these 3 lines can just use os.replace

File: saver.py (hunk 1 above)
Comment (kallewoof) [reply saver.py]: Good point, moved.

File: saver.py (hunk 2 above)
Comment (kallewoof) [reply saver.py]: Right, thanks.

File: saver.py [hunk 3]
```diff
@@ -181,5 +189,30 @@ def process_step(self, step):
...
 
+        pending_save_best_loss = os.path.exists(os.path.join(self.save_root, ".pending_save_best_loss"))
+        if pending_save_best_loss:
+            self.save_model('best_loss')
+            if is_main_process():
+                if self.old_best is not None:
+                    print(f'New best loss: {self.best_loss} from {self.old_best} (Δ{self.old_best - self.best_loss} = {100 * (self.old_best - self.best_loss) / self.old_best:.2f}%)')
+                else:
+                    print(f'New best loss: {self.best_loss}')
+                os.replace(os.path.join(self.save_root, 'best_loss'), os.path.join(self.save_root, '.pending_save_best_loss'))
```
Comment (tdrussell) [inline saver.py:200]: I think this line got messed up, it should be os.replace(os.path.join(self.save_root, '.pending_save_best_loss'), os.path.join(self.save_root, 'best_loss.txt')) right?

Comment (kallewoof) [reply saver.py:200]: Crap. Yeah. Sorry, sloppy.
